File: strings/min_add_to_make_valid_parentheses.rb
# @param {String} s
# @return {Integer}
def min_add_to_make_valid(s)
    open, invalid_close = 0, 0

    s.each_char do |paren|
        if paren == "("
            open += 1
        else
            open > 0 ? open -= 1 : invalid_close += 1
        end
    end

    invalid_close + open
end

# Faster ruby specific version
# @param {String} s
# @return {Integer}
def min_add_to_make_valid(s)
    while s.include?("()")
        s.gsub!("()", "")
    end

    return s.size
end

# 921. Minimum Add to Make Parentheses Valid
# https://leetcode.com/problems/minimum-add-to-make-parentheses-valid/

File: strings/min_add_to_make_valid_parentheses.py
# Approach 1: Stack
class Solution:
    def minAddToMakeValid(self, S: str) -> int:
        l, r = 0, 0

        for c in S:
            if c == '(':    l += 1
            elif l > 0:     l -= 1
            else:           r += 1

        return l + r


# Faster ruby specific version
# @param {String} s
# @return {Integer}
class Solution:
    def minAddToMakeValid(self, S: str) -> int:
        while '()' in S: S = S.replace('()', '')

        return len(S)


# 921. Minimum Add to Make Parentheses Valid
# https://leetcode.com/problems/minimum-add-to-make-parentheses-valid/
